### PR TITLE
[ONNX] Add runtime type checking to `export`

### PR DESCRIPTION
**Pull request by justinchuby**
This PR adds an internal wrapper on the [beartype](https://github.com/beartype/beartype) library to perform runtime type checking in `torch.onnx`. It uses beartype when it is found in the environment and is reduced to a no-op when beartype is not found.

Setting the env var `TORCH_ONNX_EXPERIMENTAL_RUNTIME_TYPE_CHECK=ERRORS` will turn on the feature. setting `TORCH_ONNX_EXPERIMENTAL_RUNTIME_TYPE_CHECK=DISABLED` will disable all checks. When not set and `beartype` is installed, a warning message is emitted.

Now when users call an api with invalid arguments e.g.

```python
torch.onnx.export(conv, y, path, export_params=True, training=False)

# traning should take TrainingModel, not bool
```

they get

```
Traceback (most recent call last):
  File "bisect_m1_error.py", line 63, in <module>
    main()
  File "bisect_m1_error.py", line 59, in main
    reveal_error()
  File "bisect_m1_error.py", line 32, in reveal_error
    torch.onnx.export(conv, y, cpu_model_path, export_params=True, training=False)
  File "<@beartype(torch.onnx.utils.export) at 0x1281f5a60>", line 136, in export
  File "pytorch/venv/lib/python3.9/site-packages/beartype/_decor/_error/errormain.py", line 301, in raise_pep_call_exception
    raise exception_cls(  # type: ignore[misc]
beartype.roar.BeartypeCallHintParamViolation: @beartyped export() parameter training=False violates type hint <class 'torch._C._onnx.TrainingMode'>, as False not instance of <protocol "torch._C._onnx.TrainingMode">.
```

when `TORCH_ONNX_EXPERIMENTAL_RUNTIME_TYPE_CHECK` is not set and `beartype` is installed, a warning message is emitted.

```
>>> torch.onnx.export("foo", "bar", "f")
<stdin>:1: CallHintViolationWarning: Traceback (most recent call last):
  File "/home/justinchu/dev/pytorch/torch/onnx/_internal/_beartype.py", line 54, in _coerce_beartype_exceptions_to_warnings
    return beartyped(*args, **kwargs)
  File "<@beartype(torch.onnx.utils.export) at 0x7f1d4ab35280>", line 39, in export
  File "/home/justinchu/anaconda3/envs/pytorch/lib/python3.9/site-packages/beartype/_decor/_error/errormain.py", line 301, in raise_pep_call_exception
    raise exception_cls(  # type: ignore[misc]
beartype.roar.BeartypeCallHintParamViolation: @beartyped export() parameter model='foo' violates type hint typing.Union[torch.nn.modules.module.Module, torch.jit._script.ScriptModule, torch.jit.ScriptFunction], as 'foo' not <protocol "torch.jit.ScriptFunction">, <protocol "torch.nn.modules.module.Module">, or <protocol "torch.jit._script.ScriptModule">.

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/justinchu/dev/pytorch/torch/onnx/_internal/_beartype.py", line 63, in _coerce_beartype_exceptions_to_warnings
    return func(*args, **kwargs)
  File "/home/justinchu/dev/pytorch/torch/onnx/utils.py", line 482, in export
    _export(
  File "/home/justinchu/dev/pytorch/torch/onnx/utils.py", line 1422, in _export
    with exporter_context(model, training, verbose):
  File "/home/justinchu/anaconda3/envs/pytorch/lib/python3.9/contextlib.py", line 119, in __enter__
    return next(self.gen)
  File "/home/justinchu/dev/pytorch/torch/onnx/utils.py", line 177, in exporter_context
    with select_model_mode_for_export(
  File "/home/justinchu/anaconda3/envs/pytorch/lib/python3.9/contextlib.py", line 119, in __enter__
    return next(self.gen)
  File "/home/justinchu/dev/pytorch/torch/onnx/utils.py", line 95, in select_model_mode_for_export
    originally_training = model.training
AttributeError: 'str' object has no attribute 'training'
```

We see the error is caught right when the type mismatch happens, improving from what otherwise would become `AttributeError: 'str' object has no attribute 'training'`

### CODE REVIEW

**Comment by justinchuby** [comment]
With this we found

```
――――――――――――――――――――――――――― TestONNXOpset.test_topk ――――――――――――――――――――――――――――
[gw0] linux -- Python 3.7.13 /opt/conda/bin/python
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/onnx/test_onnx_opset.py", line 121, in test_topk
    check_onnx_opsets_operator(module, [x, k], ops, opset_versions=[10])
  File "/var/lib/jenkins/workspace/test/onnx/test_onnx_opset.py", line 67, in check_onnx_opsets_operator
    dynamic_axes=dynamic_axes,
  File "<@beartype(torch.onnx.utils.export) at 0x7fb281514680>", line 64, in export
  File "/var/lib/jenkins/.local/lib/python3.7/site-packages/beartype/_decor/_error/errormain.py", line 302, in raise_pep_call_exception
    f'{exception_prefix}violates type hint {repr(hint)}, as '
beartype.roar.BeartypeCallHintParamViolation: @beartyped export() parameter args=[tensor([1., 2., 3., 4., 5.], requires_grad=True), tensor(3)] violates type hint typing.Union[typing.Tuple[typing.Any, ...], torch.Tensor], as [tensor([1., 2., 3., 4., 5.], requires_grad=True), tensor(3)] not tuple or <class "torch.Tensor">.
```

**Comment by justinchuby** [comment]
Supports warning thanks to discussion in https://github.com/beartype/beartype/issues/159

File: torch/onnx/_globals.py [hunk 1]
```diff
@@ -31,6 +31,13 @@ def __init__(self):
         self.operator_export_type: Optional[_C_onnx.OperatorExportTypes] = None
         self.onnx_shape_inference: bool = True
 
+        # Internal feature flags
+        self.runtime_type_check: Optional[bool] = None
```
Comment (BowenBao) [inline torch/onnx/_globals.py:35]: Would an enum be better if value `None` represents nontrivial behavior?

File: torch/onnx/_internal/_beartype.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                        return beartyped(*args, **kwargs)
+                    except _roar.BeartypeCallHintViolation:
+                        # Fall back to the original function if the beartype hint is violated.
+                        warnings.warn(
+                            traceback.format_exc(),
+                            category=errors.CallHintViolationWarning,
+                            stacklevel=2,
+                        )
+                    finally:
+                        return func(*args, **kwargs)  # type: ignore[operator] # noqa: B012
```
Comment (BowenBao) [inline torch/onnx/_internal/_beartype.py:70]: Putting "return" back to annotations?

File: torch/onnx/_internal/_beartype.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                return _coerce_beartype_exceptions_to_warnings  # type: ignore[return-value]
+
+    except ImportError:
+        # Beartype is not installed.
+        _wrapped_beartype = _no_op_decorator
+
+# Make sure that the beartype decorator is enabled whichever path we took.
+assert _wrapped_beartype is not None
+# Force cast to Callable to avoid mypy error.
+beartype = typing.cast(Callable[[_T], _T], _wrapped_beartype)
```
Comment (BowenBao) [inline torch/onnx/_internal/_beartype.py:81]: Just for my curiosity, why is this needed?

File: torch/onnx/utils.py [hunk 4]
```diff
@@ -183,6 +184,7 @@ def exporter_context(model, mode, verbose):
         yield (mode_ctx, apex_ctx, log_ctx)
 
 
+@_beartype.beartype
```
Comment (BowenBao) [inline torch/onnx/utils.py:187]: Is there a way to apply to all functions/methods under a certain module?

File: test/onnx/test_onnx_opset.py [hunk 5]
```diff
@@ -118,7 +118,7 @@ def forward(self, input, k):
         x = torch.arange(1.0, 6.0, requires_grad=True)
         k = torch.tensor(3)
         module = MyModuleDynamic()
-        check_onnx_opsets_operator(module, [x, k], ops, opset_versions=[10])
+        check_onnx_opsets_operator(module, (x, k), ops, opset_versions=[10])
```
Comment (BowenBao) [inline test/onnx/test_onnx_opset.py:121]: 🤣

**Review by BowenBao — COMMENTED** [review]
Fantastic!

File: .jenkins/caffe2/test.sh [hunk 6]
```diff
@@ -173,7 +173,7 @@ fi
 ##############
 if [[ "$BUILD_ENVIRONMENT" == *onnx* ]]; then
   pip install -q --user --no-use-pep517 "git+https://github.com/pytorch/vision.git@$(cat .github/ci_commit_pins/vision.txt)"
-  pip install -q --user ninja flatbuffers==2.0 numpy==1.21.5 onnxruntime==1.12.1
+  pip install -q --user ninja flatbuffers==2.0 numpy==1.21.5 onnxruntime==1.12.1 beartype
```
Comment (BowenBao) [inline .jenkins/caffe2/test.sh:176]: Should probably set to a fixed version for more stable CI.

File: torch/onnx/_globals.py (hunk 1 above)
Comment (justinchuby) [reply torch/onnx/_globals.py]: Done

File: torch/onnx/_internal/_beartype.py (hunk 2 above)
Comment (justinchuby) [reply torch/onnx/_internal/_beartype.py]: Done

File: torch/onnx/_internal/_beartype.py (hunk 3 above)
Comment (justinchuby) [reply torch/onnx/_internal/_beartype.py]: It's not. But I can't seem to make mypy happy with warnings enabled. 😞

File: torch/onnx/utils.py (hunk 4 above)
Comment (justinchuby) [reply torch/onnx/utils.py:187]: There should be. I saw a prototype in beartype's repo https://github.com/beartype/beartype/issues/114

File: torch/onnx/_exporter_states.py [hunk 7]
```diff
@@ -31,3 +32,14 @@ def __init__(self, params_dict, env, cur_node, onnx_block):
...
         self.onnx_block: _C.Block = onnx_block
+
+
+class RuntimeTypeCheckState(enum.Enum):
+    """Runtime type check state."""
+
+    # Runtime type checking is disabled.
+    DISABLED = 0
+    # Runtime type checking is enabled.
+    ENABLED = 1
```
Comment (BowenBao) [inline torch/onnx/_exporter_states.py:43]: nit: I would recommend swap the order of ENABLED (should we rename it ERRORS?) and WARNINGS_ONLY (WARNINGS?)

File: torch/onnx/_globals.py [hunk 8]
```diff
@@ -31,6 +32,14 @@ def __init__(self):
         self.operator_export_type: Optional[_C_onnx.OperatorExportTypes] = None
         self.onnx_shape_inference: bool = True
 
+        # Internal feature flags
+        if os.getenv("TORCH_ONNX_EXPERIMENTAL_RUNTIME_TYPE_CHECK") == "1":
```
Comment (BowenBao) [inline torch/onnx/_globals.py:36]: nit: perhaps the environmental variable values can be set to the same as enum strings ("DISABLED", etc)?

File: torch/onnx/_internal/_beartype.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+            if "return" in func.__annotations__:
+                # Remove the return type from the func function's
+                # annotations so that the beartype decorator does not complain
+                # about the return type.
+                return_type = func.__annotations__["return"]
+                del func.__annotations__["return"]
+                beartyped = _beartype_lib.beartype(func)
+                # Restore the return type to the func function's annotations
+                func.__annotations__["return"] = return_type
```
Comment (BowenBao) [inline torch/onnx/_internal/_beartype.py:69]: Is it ok for this to be here instead of the `finally:` block?

Comment (justinchuby) [reply torch/onnx/_internal/_beartype.py]: I will need to verify

File: torch/onnx/_globals.py (hunk 8 above)
Comment (justinchuby) [reply torch/onnx/_globals.py:36]: Done

File: torch/onnx/_exporter_states.py (hunk 7 above)
Comment (justinchuby) [reply torch/onnx/_exporter_states.py]: I think DISABLED=0 makes sense to be the first one (number 0). WDYT?

File: .jenkins/caffe2/test.sh (hunk 6 above)
Comment (justinchuby) [reply .jenkins/caffe2/test.sh]: Done

File: torch/onnx/_internal/_beartype.py (hunk 3 above)
Comment (justinchuby) [reply torch/onnx/_internal/_beartype.py]: Fixed!

File: torch/onnx/_internal/_beartype.py (hunk 9 above)
Comment (justinchuby) [reply torch/onnx/_internal/_beartype.py:75]: Added a test

File: torch/onnx/_internal/_beartype.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    if _beartype_lib is None:
+        # If the beartype library is not installed, return a no-op decorator
+        if runtime_check_state == _exporter_states.RuntimeTypeCheckState.ERRORS:
+            warnings.warn(
+                "TORCH_ONNX_EXPERIMENTAL_RUNTIME_TYPE_CHECK is set to 'ERRORS', "
+                "but the beartype library is not installed. "
+                "Install beartype with `pip install beartype` to enable runtime type checking."
+            )
+        return _no_op_decorator
+    else:
```
Comment (BowenBao) [inline torch/onnx/_internal/_beartype.py:55]: nit: unnecessary `else:`

**Review by BowenBao — APPROVED** [review]
🚀

**Comment by BowenBao** [comment]
Ah, please check onnx CI failures, looks related to the return annotation stuff.

**Comment by justinchuby** [comment]
Just realized and fixed

File: torch/onnx/_internal/_beartype.py (hunk 10 above)
Comment (justinchuby) [reply torch/onnx/_internal/_beartype.py]: Done

File: torch/onnx/_internal/_beartype.py [hunk 11]
```diff
@@ -52,50 +52,46 @@ def _create_beartype_decorator(
...
-                try:
-                    return beartyped(*args, **kwargs)
-                except _roar.BeartypeCallHintParamViolation:
-                    # Fall back to the original function if the beartype hint is violated.
-                    warnings.warn(
-                        traceback.format_exc(),
-                        category=errors.CallHintViolationWarning,
-                        stacklevel=2,
-                    )
-                finally:
```
Comment (justinchuby) [inline torch/onnx/_internal/_beartype.py]: Finally runs even after try has returned. Changed to remove finally

**Comment by justinchuby** [comment]
merged 🎉 special thanks to @leycec. Thank you for your generous help!

**Comment by leycec** [comment]
**Wow!** Eternal gratitude to @justinchuby for pushing this magic forward, @BowenBao for the deep review, and the whole PyTorch DevX Team for their continued excellence. Thanks so much, everyone.
